src/v/archival/tests/service_fixture.cc

```diff
@@ -385,7 +385,7 @@ void segment_matcher<Fixture>::verify_segments(
     storage::concat_segment_reader_view v{
       segments, 0, segments.back()->size_bytes(), ss::default_priority_class()};
 
-    auto expected_size = std::reduce(
+    auto expected_size = std::accumulate(
```

Comment:
my understanding is that these will produce equivalent results, but std::accumulate is a better choice because (1) we aren't looking for parallelism with std::reduce offers, and (2) std::accumulate has easier to understand semantics. but i dunno, i was just browsing Sy's opinion here https://blog.tartanllama.xyz/accumulate-vs-reduce/

Reply:
> my understanding is that these will produce equivalent results

Yes, they should, but it seems not the case while compiling with GCC. Here is the implementation of `std::reduce` in libstdc++[^1]:


```c++
  template<typename _InputIterator, typename _Tp, typename _BinaryOperation>
    _GLIBCXX20_CONSTEXPR
    _Tp
    reduce(_InputIterator __first, _InputIterator __last, _Tp __init,
	   _BinaryOperation __binary_op)
    {
      using __ref = typename iterator_traits<_InputIterator>::reference;
      static_assert(is_invocable_r_v<_Tp, _BinaryOperation&, _Tp&, __ref>);
      static_assert(is_invocable_r_v<_Tp, _BinaryOperation&, __ref, _Tp&>);
      static_assert(is_invocable_r_v<_Tp, _BinaryOperation&, _Tp&, _Tp&>);
      static_assert(is_invocable_r_v<_Tp, _BinaryOperation&, __ref, __ref>);
      if constexpr (__is_random_access_iter<_InputIterator>::value)
	{
	  while ((__last - __first) >= 4)
	    {
	      _Tp __v1 = __binary_op(__first[0], __first[1]);
	      _Tp __v2 = __binary_op(__first[2], __first[3]);
	      _Tp __v3 = __binary_op(__v1, __v2);
	      __init = __binary_op(__init, __v3);
	      __first += 4;
	    }
	}
      for (; __first != __last; ++__first)
	__init = __binary_op(__init, *__first);
      return __init;
    }
```

It requires that the type of element, init value and return value should be the same(or convertible, roughly speaking), which differs from that of libc++.

Or do we only support clang?

[^1]: https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/std/numeric#L284

Reply:
redpanda only compiles with clang (for now), but that is primarily because current gcc has some coroutine issues. however, redpanda should compile with libc++ or libstdc++.

Reply:
what's the error you are seeing? because it looks like init and return value of the binary op are the same in this case?

Reply:
> what's the error you are seeing?

oh my fault, I forgot to paste the error message and missed some important information. 

Here is the error message:

```txt
In file included from /data/Workspace/iSoft/redpanda/build/deps_install/include/absl/algorithm/container.h:46:
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/numeric:293:7: error: static_assert failed due to requirement 'is_invocable_r_v<unsigned long, (lambda at /data/Workspace/iSoft/redpanda/src/v/archival/tests/service_fixture.cc:392:7) &, unsigned long &, unsigned long &>'
      static_assert(is_invocable_r_v<_Tp, _BinaryOperation&, _Tp&, _Tp&>);
      ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/data/Workspace/iSoft/redpanda/src/v/archival/tests/service_fixture.cc:388:31: note: in instantiation of function template specialization 'std::reduce<__gnu_cxx::__normal_iterator<seastar::lw_shared_ptr<storage::segment> *, std::vector<seastar::lw_shared_ptr<storage::segment>>>, unsigned long, (lambda at /data/Workspace/iSoft/redpanda/src/v/archival/tests/service_fixture.cc:392:7)>' requested here
    auto expected_size = std::reduce(
                              ^
/data/Workspace/iSoft/redpanda/src/v/archival/tests/service_fixture.cc:438:16: note: in instantiation of member function 'archival::segment_matcher<archival::archiver_fixture>::verify_segments' requested here
template class segment_matcher<archiver_fixture>;
               ^
/data/Workspace/iSoft/redpanda/src/v/archival/tests/service_fixture.cc:392:50: error: invalid operands to binary expression ('seastar::lw_shared_ptr<storage::segment>' and 'size_t' (aka 'unsigned long'))
      [](auto acc, const auto& seg) { return acc + seg->size_bytes(); });
                                             ~~~ ^ ~~~~~~~~~~~~~~~~~
```

> because it looks like init and return value of the binary op are the same in this case?

libstdc++ version also requires that both the parameters of the binary op should be of the element type.

```c++
static_assert(is_invocable_r_v<_Tp, _BinaryOperation&, __ref, __ref>);
```

The requirement is needed becaused the binary op can be applied directly to the elements:

```c++
	  while ((__last - __first) >= 4)
	    {
	      _Tp __v1 = __binary_op(__first[0], __first[1]);
	      _Tp __v2 = __binary_op(__first[2], __first[3]);
	      _Tp __v3 = __binary_op(__v1, __v2);
	      __init = __binary_op(__init, __v3);
	      __first += 4;
	    }
```

**So the element type, init type, return type, and types of parameters of the binary op, should be the same (or convertible).**

But in our case, `acc` is the sum, and `seg` is an element deferenced from the input iterator, which are of different types.